rust/template/distributed_datalog/src/tcp_channel/txnbuf.rs

```diff
@@ -0,0 +1,245 @@
+use std::collections::LinkedList;
+use std::fmt::Debug;
+use std::io::Write;
+use std::mem::replace;
+
+use bincode::serialize_into;
+use serde::Serialize;
+
+use crate::observe::Observer;
+use crate::tcp_channel::message::Message;
+
+/// A type representing the updates of a transaction.
+type Transaction<T> = LinkedList<Vec<T>>;
+
+/// A buffer for transactions.
+#[derive(Debug)]
+pub enum TxnBuf<W, T>
+where
+    W: Debug,
+    T: Debug,
+{
+    /// A list of buffered transactions.
+    Updates {
+        /// We merge all "completed" transactions (i.e., those for which
+        /// we have seen a commit) into a single one.
+        complete: Transaction<T>,
+        /// The transaction currently in progress.
+        ongoing: Option<Transaction<T>>,
+        /// We have received an `on_completed` event.
+        on_completed: bool,
+    },
+    /// A writer is present and we no longer need to buffer
+    /// transactions.
+    Writer(W),
+}
+
+impl<W, T> TxnBuf<W, T>
+where
+    W: Debug + Send + Write,
+    T: Debug + Send + Serialize,
+{
+    /// Convert the `TxnBuf` into the `Writer` variant.
+    ///
+    /// An error return indicates a failure to flush all buffered
+    /// transactions. The objects is in an undefined state afterwards.
+    pub fn make_writer(&mut self, mut writer: W) -> Result<(), String> {
```

Comment:
This function name refers to internal structure of `TxnBuf`, not what it does from the user's perspective. Can we give it a more intuitive name, e.g., `set_mode_passthrough()`.

Reply:
Thanks for your review! Well, it's an `enum` so the there really isn't anything internal about the structure. But the name you proposed is fine with me as well.

Reply:
As a user, it does not matter to me that it's an enum. It's just a filter that can operate in either buffered or passthrough mode. In reviewing the code, I came across a call to this function before reading the function itself, and I couldn't figure our what it was doing, hence my suggestion to rename.